### PR TITLE
VDS-1180 Create backend for cvs pricing import validation

### PR DESCRIPTION
**Pull request by trueboroda**

### CODE REVIEW

File: src/VirtoCommerce.SimpleExportImportModule.Core/ModuleConstants.cs [hunk 1]
```diff
@@ -2,10 +2,25 @@ namespace VirtoCommerce.SimpleExportImportModule.Core
 {
     public static class ModuleConstants
     {
+        public const long FileMaxSize = 1024 * 1024 * 8;
```
Comment (asvishnyakov) [inline src/VirtoCommerce.SimpleExportImportModule.Core/ModuleConstants.cs]: May we split it into multiple constants (byte, Kb, Mb)...

File: src/VirtoCommerce.SimpleExportImportModule.Core/ModuleConstants.cs [hunk 2]
```diff
@@ -2,10 +2,25 @@ namespace VirtoCommerce.SimpleExportImportModule.Core
...
+
```
Comment (asvishnyakov) [inline src/VirtoCommerce.SimpleExportImportModule.Core/ModuleConstants.cs:6]: Move to Settings please

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            {
+                errorsList.Add(ModuleConstants.ValidationErrors.FileNotExisted);
+            }
+            else if (blobInfo.Size > ModuleConstants.FileMaxSize)
+            {
+                errorsList.Add(ModuleConstants.ValidationErrors.ExceedingFileMaxSize);
+            }
+            else
+            {
+                var stream = _blobStorageProvider.OpenRead(fileUrl);
```
Comment (asvishnyakov) [inline src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: Dispose Stream

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                var csvConfiguration = new Configuration(CultureInfo.InvariantCulture) { Delimiter = ";" };
+                using var streamReader = new StreamReader(stream);
+                using var csvReader = new CsvReader(streamReader, csvConfiguration);
+
+                var headerStr = await streamReader.ReadLineAsync();
```
Comment (asvishnyakov) [inline src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: Don't use abbreviations please, name it just header or headerLine

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+                if (headerStr == null || headerStr.Trim() == "")
```
Comment (asvishnyakov) [inline src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: Why Trim is here?

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                {
+                    errorsList.Add(ModuleConstants.ValidationErrors.NoData);
+                }
+                else
+                {
+                    var headerColumns = headerStr.Split(csvConfiguration.Delimiter);
+
+                    if (headerColumns.Length < 2)
```
Comment (asvishnyakov) [inline src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: May we just use headerLine.Contains(csvConfiguraton.Delimiter)?

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                errorsList.Add(ModuleConstants.ValidationErrors.FileNotExisted);
+            }
+            else if (blobInfo.Size > ModuleConstants.FileMaxSize)
+            {
+                errorsList.Add(ModuleConstants.ValidationErrors.ExceedingFileMaxSize);
+            }
+            else
+            {
+                var stream = _blobStorageProvider.OpenRead(fileUrl);
+                var csvConfiguration = new Configuration(CultureInfo.InvariantCulture) { Delimiter = ";" };
```
Comment (asvishnyakov) [inline src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: Use https://github.com/VirtoCommerce/vc-module-simple-export-import/blob/c51dbe25d0dcd28442f8be543e59fdfd9b1d6780/src/VirtoCommerce.SimpleExportImportModule.Core/Models/ImportConfiguration.cs to hardcode delimiter in single place

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                    var headerColumns = headerStr.Split(csvConfiguration.Delimiter);
+
+                    if (headerColumns.Length < 2)
+                    {
+                        errorsList.Add(ModuleConstants.ValidationErrors.WrongDelimiter);
+                    }
+
+                    var fistDataRowStr = await streamReader.ReadLineAsync();
+
+                    if (fistDataRowStr == null || fistDataRowStr.Trim() == "")
```
Comment (asvishnyakov) [inline src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: Trim?

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+                }
+
+                if (errorsList.Count == 0)
+                {
+                    var totalCount = 0;
+
+                    while (csvReader.Read())
+                    {
+                        totalCount++;
+                    }
```
Comment (asvishnyakov) [inline src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: May we instead reuse code from data source? For example, you miss case about header: it shouldn't be counted, while in this code it is.

**Review by asvishnyakov — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs (hunk 5 above)
Comment (trueboroda) [reply src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: A string with only spaces is also considered no data.

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs (hunk 8 above)
Comment (trueboroda) [reply src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: See above

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs (hunk 9 above)
Comment (trueboroda) [reply src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: Why header should be counted?

File: src/VirtoCommerce.SimpleExportImportModule.Core/ModuleConstants.cs (hunk 2 above)
Comment (trueboroda) [reply src/VirtoCommerce.SimpleExportImportModule.Core/ModuleConstants.cs:6]: It is a separate task. https://virtocommerce.atlassian.net/browse/VDS-738

Comment (asvishnyakov) [reply src/VirtoCommerce.SimpleExportImportModule.Core/ModuleConstants.cs:6]: I mean static `Settings` class below

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs (hunk 5 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: So when we should use string.IsNullOrWhiteSpace

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs (hunk 9 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: It shouldn't and now you count it. See code of GetTotalCount from data source

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs (hunk 5 above)
Comment (trueboroda) [reply src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: Предложил бы сразу )

Comment (trueboroda) [reply src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: We don't have such extension method for string

Comment (asvishnyakov) [reply src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: https://docs.microsoft.com/en-us/dotnet/api/system.string.isnullorwhitespace?view=netcore-3.0

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+            }
+        }
+
+        private static void ReadCsvHeader(CsvReader csvReader)
+        {
+            csvReader.Read();
+            csvReader.ReadHeader();
+        }
+
+        private static void SeekStreamReaderToStart(StreamReader streamReader)
```
Comment (asvishnyakov) [inline src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs:126]: Why we have separate method for not reused code without large complexity?

File: src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+            return result;
+        }
+
+        private static void ValidateLineLimit(CsvReader csvReader, List<string> errorsList)
+        {
+            if (errorsList.Count == 0)
+            {
+                ReadCsvHeader(csvReader);
+
+                var totalCount = 1;
```
Comment (asvishnyakov) [inline src/VirtoCommerce.SimpleExportImportModule.Data/Services/CsvPriceDataValidator.cs]: Should be 0. Try to remove `+ 1` from test for exceeding lines and see what happen. It also mean we should write tests not only for invalid csv ("positive scenarios"), but also for valid csv ("negative scenarios")

**Review by asvishnyakov — CHANGES_REQUESTED** [review]

**Review by asvishnyakov — APPROVED** [review]
